c2rust-analyze/src/borrowck/mod.rs

```diff
@@ -15,25 +19,98 @@ mod dump;
 mod type_check;
 
 #[derive(Clone, Copy, PartialEq, Eq, Debug, Hash, Default)]
-pub struct Label {
+pub struct Label<'tcx> {
+    /// The [`Origin`] of this type
     pub origin: Option<Origin>,
+    /// The [Origins](`Origin`) associated with each lifetime
+    /// parameter of this type, if applicable
+    pub origin_params: Option<&'tcx [(OriginKind<'tcx>, Origin)]>,
     pub perm: PermissionSet,
 }
 
-pub type LTy<'tcx> = LabeledTy<'tcx, Label>;
-pub type LTyCtxt<'tcx> = LabeledTyCtxt<'tcx, Label>;
+pub type LTy<'tcx> = LabeledTy<'tcx, Label<'tcx>>;
+pub type LTyCtxt<'tcx> = LabeledTyCtxt<'tcx, Label<'tcx>>;
+
+/// Metadata describing lifetimes and lifetime parameters
+/// of a struct field.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FieldMetadata<'tcx> {
+    /// The lifetime of the field, e.g. `*mut &'a mut foo_type`
+    /// would have an index set of {'h0, 'a}
+    pub lifetime: IndexSet<OriginKind<'tcx>>,
+    /// The lifetime parameters of a field, e.g. if a struct
+    /// `foo<'a, 'b>` is a field of `bar<'c, 'd>` as field: `foo<'c, 'd>`,
+    /// the lifetime params would be a set {'c, 'd}
+    pub lifetime_params: IndexSet<OriginKind<'tcx>>,
+    /// The type of the field when fully dereferenced, e.g.
+    /// `&mut &mut foo_type` would have a type of `foo_type`
+    pub fully_derefed_ty: Option<Ty<'tcx>>,
+}
+
+impl Default for FieldMetadata<'_> {
+    fn default() -> Self {
+        Self {
+            lifetime: IndexSet::new(),
+            lifetime_params: IndexSet::new(),
+            fully_derefed_ty: None,
+        }
+    }
+}
+
+/// Metadata describing the lifetime parameters and fields
+/// of a struct.
+#[derive(Clone, PartialEq, Eq)]
+pub struct AdtMetadata<'tcx> {
+    /// The lifetime parameters of a structure, including
+    /// hypothetical lifetimes derived from pointer fields.
+    pub lifetime_params: IndexSet<OriginKind<'tcx>>,
+    pub field_info: IndexMap<DefId, FieldMetadata<'tcx>>,
+}
+
+impl Default for AdtMetadata<'_> {
+    fn default() -> Self {
+        Self {
+            lifetime_params: IndexSet::new(),
+            field_info: IndexMap::new(),
+        }
+    }
+}
+
+/// An origin parameter to resolve in a MIR body
+/// that will get mapped to a concrete Origin to
+/// provide to polonius.
+#[derive(Hash, PartialEq, Eq, Clone, Copy)]
+pub enum OriginKind<'tcx> {
+    /// An existing region, i.e. `'a` in `&'a foo`
+    Actual(Region<'tcx>),
+    /// A hypothesized region derived from a pointer type
+    /// e.g. `'h0` derived from the pointer in `*mut foo`
+    Hypothetical(i64),
+}
```

Comment:
I think this type is used in two different ways (let me know if I have this right):

1. In `AdtMetadata` for listing all the (real and hypothetical) origin parameters of a type.  In that case, `Actual` is only used with `ReEarlyBound`.
2. In `FieldMetadata` for listing the origin parameters of the field type.  There, `Actual` can either be `ReEarlyBound` or `ReStatic`, and both the `ReEarlyBound` and `Hypothetical` cases can only refer to the parameters listed in the parent `AdtMetadata`.

I would be in favor of splitting this into two separate types, maybe named `OriginParam` for use 1 and `OriginArg` for use 2.  I think this let us be clearer in docs and type signatures about whether we're talking about a declaration or a use of an origin parameter and which `RegionKind` variants could reasonably appear there (in fact, `OriginParam` could probably use `EarlyBoundRegion` directly, if `ReEarlyBound` is really the only legal variant to use with `Actual`).

Reply:
Let me know if 8f5fec6c accomplishes this

Reply:
That looks about right.  I have a few more comments on parts of it, which I'll add separately.